### PR TITLE
Add Notebook Save integration tests

### PR DESCRIPTION
**Pull request by kevcunnane**
Wrote test to verify save behavior
Fixed issues found during testing,
specifically around how we notify dirty state change to extensions

### CODE REVIEW

File: extensions/integration-tests/src/notebook.test.ts [hunk 1]
```diff
@@ -180,6 +184,41 @@ class NotebookTester {
...
+		assert(notebook.document.kernelSpec.name === 'SQL', `Expected first kernel name: SQL, Actual: ${notebook.document.kernelSpec.name}`);
+		assert(notebook.document.isDirty === true, 'Notebook should be dirty after edit');
+
+		// When I save it, it should no longer be dirty
+		let saved = await notebook.document.save();
+		assert(saved === true, 'Expect save to succeed');
+		// Note: need to sleep after save as the change events happen after save
+		// We need to give back the thread or the event won't have been drained.
+		// This is consistent with VSCode APIs, so keeping as-is
+		await sleep(10);
```
Comment (alanrenmsft) [inline extensions/integration-tests/src/notebook.test.ts]: is 10ms reliable enough?

File: extensions/integration-tests/src/notebook.test.ts [hunk 2]
```diff
@@ -180,6 +184,41 @@ class NotebookTester {
...
+		// And when I edit again, should become dirty
+		let edited = await notebook.edit(builder => {
+			builder.insertCell({
+				cell_type: CellTypes.Code,
+				source: ''
+			});
+		});
+		assert(edited === true, 'Expect edit to succeed');
+		await sleep(10);
+		assert(notebook.document.isDirty === true, 'Notebook should be dirty after edit');
```
Comment (alanrenmsft) [inline extensions/integration-tests/src/notebook.test.ts]: from the log message, you can't tell whether this is first save or the second, same applies to all the 2nd round asserts.

Comment (kevcunnane) [reply extensions/integration-tests/src/notebook.test.ts]: Fair point. I can change but usually I'd tend to rely on line numbers for which failed?

File: extensions/integration-tests/src/notebook.test.ts (hunk 1 above)
Comment (kevcunnane) [reply extensions/integration-tests/src/notebook.test.ts]: It should be, I had it higher but I'm concerned about making integration tests very long. These events come right after each other so mostly ceding the thread for any length of time should do it.

**Pull request merged**